cdc/log.cc

```diff
@@ -526,36 +551,173 @@ class transformer final {
                     r.for_each_cell([&](column_id id, const atomic_cell_or_collection& cell) {
                         auto& cdef = _schema->column_at(ckind, id);
                         auto* dst = _log_schema->get_column_definition(to_bytes("_" + cdef.name()));
-                        // TODO: collections.
-                        if (cdef.is_atomic()) {
-                            column_op op;
+                        auto res_type = static_pointer_cast<const tuple_type_impl>(dst->type)->type(1);
+                        column_op op = column_op::del;
+                        values[1] = values[2] = std::nullopt;
+                        auto has_pirow = pirow && pirow->has(cdef.name_as_text());
 
-                            values[1] = values[2] = std::nullopt;
+                        if (cdef.is_atomic()) {
                             auto view = cell.as_atomic_cell(cdef);
                             if (view.is_live()) {
                                 op = column_op::set;
                                 values[1] = view.value().linearize();
                                 if (view.is_live_and_has_ttl()) {
                                     values[2] = long_type->decompose(data_value(view.ttl().count()));
                                 }
-                            } else {
-                                op = column_op::del;
-                            }
+                            } 
+                        } else {
+                            auto mv = cell.as_collection_mutation();
+                            auto in = collection_mutation_input_stream(mv.data);
+                            auto tomb = tombstone{};
+                            auto now = gc_clock::time_point::min();
+
+                            std::optional<column_op> oop;
+                            op = column_op::add;
+                            std::vector<bytes> buf;
+
+                            auto process = [&](auto key_2_type, auto append_value) {
```

Comment:
Space between: `]` and `(`. Gratuitous `auto`. Knowing these types at read-time would actually help understand the code.

Reply:
No space. You are delusional. :-)
These are lambda callbacks, so types are tricky (don't want to make them std::func here). But maybe better param-names.

Reply:
Redone somewhat.